File: web/readiness.js
"use strict";
let readinessPending=false;
const readinessExpanded=new Set();
Object.assign(titles,{readiness:['Operational readiness','Prepare one supervised pilot. Every gate has an owner and a reason.']});

function openReadiness() {
 view='readiness';selected=null;
 document.querySelectorAll('[data-view]').forEach(b=>{b.removeAttribute('aria-current');if(b.dataset.view===view)b.setAttribute('aria-current','page');});
 render();window.scrollTo(0,0);
}

function readinessSummary(root) {
 const data=state.readiness;if(!data)return;
 const panel=el('div',null,'readiness-summary');
 panel.append(el('p','OPERATING POSTURE','eyebrow'),el('h2',data.activeWorkers?'Existing work needs supervision':data.pilotPassed?'Review the next approved scope':'Prepare the first supervised pilot'),
  el('p',data.nextAction),button('Review readiness & blockers',openReadiness));
 root.append(panel);
}

async function runReadiness(operation) {
 if(readinessPending||!connected)return;
 readinessPending=true;render();
 try{
  await api('/api/readiness',{method:'POST',headers:{'Content-Type':'application/json','X-CSRF-Token':csrf},body:JSON.stringify({operation})});
  showNotice(operation==='inspect'?'Local metadata checks started. No task or model request.':'Synthetic lifecycle rehearsal started. This cannot certify a real pilot.');
  await refresh();
 }catch(error){showNotice(error.message,true);}
 finally{readinessPending=false;render();}
}

function readiness(root) {
 const d=state.readiness,m=state.meta;if(!d){root.append(empty('Readiness unavailable','Refresh the local connection.'));return;}
 const controls=el('div',null,'observation-toolbar');
 const inspect=button('Inspect local readiness',()=>runReadiness('inspect'),'primary');
 const rehearse=button('Run isolated rehearsal',()=>runReadiness('rehearse'));
 inspect.disabled=rehearse.disabled=readinessPending||d.job?.status==='running';controls.append(inspect,rehearse);root.append(controls);
 if(d.job?.status==='running')root.append(el('p','Running '+d.job.operation+'… No live dispatch or acceptance commands.','metric-note'));
 if(d.job?.status==='complete')root.append(el('p',(d.job.operation==='inspect'?'Local readiness inspection completed':'Synthetic rehearsal completed')+' · '+when(d.job.finishedAt)+'. Controller authority is unchanged.','metric-note'));
 if(d.job?.status==='failed')root.append(callout('Check did not complete',d.job.error));
 root.append(el('p','Local checks: '+when(d.localObservedAt)+' · Native inventory: '+when(d.nativeObservedAt)+'. Observations expire after 15 minutes; packet preflight expires after 5.','metric-note'));
 root.append(callout(d.activeWorkers?'Supervision required':m.paused?'Dispatch parked safely':'Brain verification required',d.nextAction));
 root.append(section('Before the platform drives','A green diagnostic never authorizes work.'));
 root.append(table(['Gate','Recorded state','Next action'],[
  ['Designated brain',textCell(d.brainObservedStatus,d.brainId||'Not configured'),d.brainFresh?'Recent ledger checkpoint; recheck native state at dispatch.':'Open the designated brain and perform a read-only reconciliation.'],
  ['Approved scope',state.queue.filter(q=>q.status==='approved').length+' approved / '+state.queue.length+' prepared','Review one immutable inheritance seed, then approve its exact hashes.'],
  ['Native mappings',d.repositories.filter(r=>r.mappingObserved).length+' / '+d.repositories.length+' observed','Register existing checkouts in Codex; then the brain records exact native IDs.'],
  ['Dispatch',m.paused?'Paused':'Enabled in ledger','Resume is a separate request processed by the brain, never by this diagnostic.'],
  ['Heartbeat',d.heartbeat.status+' (recorded)','A paused heartbeat does not wake from this webpage. Start with the brain handoff below.'],
  ['Real worker pilot',d.pilotPassed?'Recorded accepted':'Not accepted · one worker limit','Complete and independently verify a real approved worker; simulation does not qualify.'],
  ['Pending controls',d.pendingControls+' pending / '+d.uncertainControls+' uncertain','Reconcile uncertain native actions before retrying.']
 ]));
 root.append(section('Repository preparation','Metadata checks do not prove safe setup, CI or acceptance.'));
 root.append(table(['Repository / policy','Checkout & native mapping','Required follow-through'],d.repositories.map(r=>{
  const detail=el('details',null,'readiness-detail');detail.append(el('summary',r.issues.length?r.issues.length+' setup issue(s)':'Registration observed; review execution gates'));
  detail.open=readinessExpanded.has(r.repository);
  detail.addEventListener('toggle',()=>{if(detail.isConnected){if(detail.open)readinessExpanded.add(r.repository);else readinessExpanded.delete(r.repository);}});
  const list=el('ul');r.issues.forEach(v=>list.append(el('li',v)));list.append(el('li',r.setupReview),el('li',r.ciNote));
  if(r.candidateProjectId&&!r.configuredProjectId)list.append(el('li','Observed matching project: '+r.candidateProjectId+'. Mapping has not been applied.'));
  detail.append(list);return[textCell(r.repository,r.profile+' · '+r.mergePolicy+' merge'),textCell(r.commit?.slice(0,12)||'Not observed',r.mappingObserved?'Native mapping observed':'Native mapping unverified'),detail];
 })));
 root.append(section('Packet launch preview','Read-only explanation using the same packet gates as dispatch.'));
 if(!d.packets.length)root.append(el('p','No packet has been prepared. A roadmap checkbox is not an executable queue item. Select one bounded scope and ask the brain to prepare its inheritance and evidence contract.','muted'));
 for(const p of d.packets){const sectionNode=el('section',null,'detail');sectionNode.append(el('h3',p.packetId+' · '+p.repository),badge(p.status==='complete'?'complete':p.status==='dispatched'?'owned':p.ledgerEligible?'ledger gates met':'blocked'),el('p','Seed '+p.seedHash.slice(0,12),'mono'));
  const list=el('ul');p.issues.forEach(i=>list.append(el('li',i.detail)));sectionNode.append(list);if(p.ledgerEligible)sectionNode.append(el('p','Still requires a live brain cycle and current external checks. This preview does not create a task.','muted'));root.append(sectionNode);}
 root.append(section('Safe handoff to the brain','Copy this into the designated Codex task. It requests inspection, not approval.'));
 root.append(el('pre',d.brainPrompt));const copy=button('Copy brain handoff',async()=>{try{await navigator.clipboard.writeText(d.brainPrompt);showNotice('Read-only handoff copied. Paste it into the designated brain task.');}catch{showNotice('Clipboard unavailable. Copy the displayed text manually.',true);}});root.append(copy,el('p','This dashboard cannot open or wake a native task by itself. Queued control requests execute only when the brain is active.','metric-note'));
 root.append(section('Isolated lifecycle rehearsal','No native task, product checkout, inference request or acceptance process.'));
 if(d.rehearsal){root.append(el('p',d.rehearsal.status+' · '+when(d.rehearsal.at)+' · Synthetic only; real pilot remains '+(d.pilotPassed?'separately recorded':'unaccepted'), 'metric-note'));
  root.append(table(['Simulated safety check','Result'],d.rehearsal.steps.map(s=>[s.check,s.status])));
  if(d.rehearsal.artifactId)root.append(button('Read rehearsal evidence',()=>{selected=d.rehearsal.artifactId;render();}));
 }else root.append(el('p','Not run. Use the rehearsal to check controller state transitions before a real pilot.','muted'));
 root.append(el('p',d.limits,'metric-note'));
 if(selected)artifactReader(root,selected);
}

File: orchestrator/readiness.py
"""Deterministic operating diagnostics. Never authorizes or dispatches work."""
from __future__ import annotations

from pathlib import Path
import subprocess
import time

from .core import ACTIVE, digest, eligibility_issues, require
from .observations import save

FRESH_SECONDS = 15 * 60


def native_observation(ledger, record):
    """Import a bounded operator observation from native Codex tools, not private APIs."""
    require(isinstance(record, dict) and set(record) == {"schemaVersion", "observedAt", "brain", "projects"}, "Invalid native observation")
    require(record["schemaVersion"] == 1 and type(record["observedAt"]) in (int, float)
        and 0 <= time.time() - record["observedAt"] <= 300, "Native observation must be current (five minutes)")
    brain = record["brain"]
    require(isinstance(brain, dict) and {"id", "status"} <= set(brain) <= {"id", "status", "title"}, "Invalid brain observation")
    if "title" in brain:
        require(isinstance(brain["title"], str) and 0 < len(brain["title"]) <= 200
            and all(ord(c) >= 32 for c in brain["title"]), "Invalid brain title")
    state = ledger.snapshot()
    require(isinstance(brain["id"], str) and bool(brain["id"]) and brain["id"] == state["meta"]["brainId"]
        and brain["status"] in ("idle", "running", "unavailable", "unknown"), "Native brain observation does not match portfolio")
    projects = record["projects"]
    require(isinstance(projects, list) and len(projects) <= 500, "Bounded native project inventory required")
    paths = {str(Path(r["path"]).resolve()) for r in state["repositories"] if r["path"]}
    selected, identities = [], set()
    for p in projects:
        require(isinstance(p, dict) and set(p) == {"projectId", "path", "hostId", "isGitRepository"}, "Invalid native project entry")
        require(all(isinstance(p[k], str) and 0 < len(p[k]) <= 4096 for k in ("projectId", "path", "hostId"))
            and type(p["isGitRepository"]) is bool and Path(p["path"]).is_absolute(), "Invalid native project values")
        require(p["projectId"] not in identities, "Duplicate native project identity")
        identities.add(p["projectId"])
        if str(Path(p["path"]).resolve()) in paths:
            selected.append(p)
    retained = {**record, "projects": selected, "source": "operator_recorded_native_tools", "authorization": False}
    with ledger.tx() as db:
        save(db, "native", retained)
    return {"status": "recorded", "matchedProjects": len(selected), "observedAt": record["observedAt"]}


def inspect_repository(repo):
    row = {"repository": repo["id"], "configurationHash": digest(repo), "status": "unavailable", "commit": None,
        "reason": None, "hasAgents": False, "workflowCount": None, "setupReview": "not_verified"}
    if not repo["path"] or not Path(repo["path"]).is_dir():
        return {**row, "reason": "Configured checkout is absent; do not bootstrap it implicitly."}
    root = Path(repo["path"]).resolve()
    try:
        def git(*args):
            return subprocess.run(["git", "-C", str(root), *args], capture_output=True, text=True,
                timeout=10, check=True).stdout.strip()
        require(Path(git("rev-parse", "--show-toplevel")).resolve() == root, "Configured path is not the exact repository root")
        commit = git("rev-parse", "--verify", "--end-of-options", repo["ref"] + "^{commit}")
        agents = root / "AGENTS.md"
        workflows = root / ".github" / "workflows"
        count = None if workflows.is_symlink() or not workflows.resolve().is_relative_to(root) else sum(p.is_file() and not p.is_symlink() for pattern in ("*.yml", "*.yaml") for p in workflows.glob(pattern))
        row.update(status="observed", commit=commit, hasAgents=agents.is_file() and not agents.is_symlink(), workflowCount=count,
            worktreeIncludePresent=(root / ".worktreeinclude").exists(), localEnvironmentPresent=(root / ".codex" / "environments").exists())
    except (OSError, subprocess.SubprocessError, ValueError):
        row["reason"] = "Cannot verify the exact configured Git root/ref. Review its path and commit before dispatch."
    return row


def collect(ledger):
    state = ledger.snapshot()
    record = {"at": time.time(), "repositories": [inspect_repository(r) for r in state["repositories"]],
        "method": "Local Git metadata only. No fetch, repository code, acceptance, model request or native action."}
    with ledger.tx() as db:
        save(db, "readiness", record)
    return {"status": "complete", "repositories": len(record["repositories"]), "observedAt": record["at"]}


def diagnose(ledger, state=None):
    state = state or ledger.snapshot()
    meta, obs, now = state["meta"], state["observations"], time.time()
    admission = state.get("admission", {})
    local, native = obs.get("readiness") or {}, obs.get("native") or {}
    def fresh(at):
        return bool(at) and 0 <= now - at <= FRESH_SECONDS
    local_fresh, native_fresh = fresh(local.get("at")), fresh(native.get("observedAt"))
    active = [w for w in state["workers"] if w["status"] in ACTIVE]
    rows = []
    for repo in state["repositories"]:
        measured = next((r for r in local.get("repositories", []) if r["repository"] == repo["id"]), {})
        if measured.get("configurationHash") != digest(repo): measured = {}
        candidates = [p for p in native.get("projects", []) if repo["path"] and
            Path(p["path"]).resolve() == Path(repo["path"]).resolve() and p["hostId"] == "local" and p["isGitRepository"]]
        matched = native_fresh and any(p["projectId"] == repo["projectId"] for p in candidates)
        issues = []
        if not local_fresh: issues.append("Refresh local readiness checks; measurements are missing or stale.")
        if measured.get("status") != "observed": issues.append(measured.get("reason") or "Checkout/ref has not been inspected.")
        if not native_fresh: issues.append("Ask the brain to refresh native project inventory; the webpage cannot call native tools.")
        if not matched: issues.append("Register the existing checkout in Codex and record its exact project mapping; no mapping is inferred.")
        if not measured.get("hasAgents"): issues.append("No regular AGENTS.md observed; review the repository's execution contract.")
        rows.append({"repository": repo["id"], "profile": repo["policyProfile"], "mergePolicy": repo["mergePolicy"],
            "configuredProjectId": repo["projectId"], "candidateProjectId": candidates[0]["projectId"] if native_fresh and len(candidates) == 1 else None,
            "mappingObserved": bool(matched), "commit": measured.get("commit"), "workflowCount": measured.get("workflowCount"),
            "registrationReady": not issues, "issues": issues,
            "setupReview": "Review native setup scripts and ignored-file copying before each new worktree. Absence of repository settings is not proof of safe app settings.",
            "ciNote": "Workflow presence does not prove required checks or runner availability." if measured.get("workflowCount") else "No repository workflow observed. Define a valid CI evidence route before approving a pilot; local tests are not automatically CI."})
    packets = []
    for q in state["queue"]:
        repo = next(r for r in state["repositories"] if r["id"] == q["repository"])
        if q["status"] in ("dispatched", "complete"):
            packets.append({"queueId": q["id"], "repository": q["repository"], "packetId": q["packetId"], "seedHash": q["seedHash"],
                "status": q["status"], "ledgerEligible": False, "issues": [{"code": "complete", "detail": "Completed under its recorded evidence contract; no new launch. Merge and archive remain separate."}] if q["status"] == "complete" else
                    [{"code": "owned", "detail": "Already dispatched; reconcile its existing owner, never create a duplicate."}]})
            continue
        issues = eligibility_issues(q, repo, ledger.document(q["seedHash"]), state["workers"], now)
        if admission.get("dispatchBlocked"):
            issues.append({"code": "platform_enrollment", "detail": admission["reason"]})
        if meta["paused"]: issues.append({"code": "paused", "detail": "Dispatch paused; any resume request must be processed by the designated brain."})
        if len(active) >= meta["concurrency"]: issues.append({"code": "capacity", "detail": "Worker capacity is reserved."})
        if any(w["repository"] == q["repository"] for w in active): issues.append({"code": "repository_owned", "detail": "Repository is owned by an existing worker."})
        packets.append({"queueId": q["id"], "repository": q["repository"], "packetId": q["packetId"], "seedHash": q["seedHash"], "status": q["status"], "ledgerEligible": not issues, "issues": issues})
    pending = [c for c in state["commands"] if c["status"] in ("queued", "processing")]
    status = "supervision_required" if active else "parked" if not any(q["status"] == "approved" for q in state["queue"]) else "brain_check_required"
    return {"status": status, "checkedAt": now, "readOnly": True, "launchAuthorized": False,
        "brainId": meta["brainId"], "brainObservedStatus": native.get("brain", {}).get("status", "unknown") if native_fresh else "unknown",
        "lastReconciled": meta["lastReconciled"], "brainFresh": bool(meta["lastReconciled"]) and 0 <= now - meta["lastReconciled"] <= 1800,
        "localObservedAt": local.get("at"), "nativeObservedAt": native.get("observedAt"), "localFresh": local_fresh, "nativeFresh": native_fresh,
        "repositories": rows, "packets": packets, "pendingControls": len(pending), "uncertainControls": sum(c["status"] == "processing" for c in pending),
        "activeWorkers": len(active), "pilotPassed": meta["pilotPassed"], "concurrency": meta["concurrency"],
        "heartbeat": {**meta["heartbeat"], "source": "recorded_ledger_state_not_live_poll"}, "rehearsal": obs.get("rehearsal"),
        "admission": admission,
        "nextAction": admission["reason"] if admission.get("dispatchBlocked") else "Reconcile existing ownership and pending native actions." if active or pending else
            "Ask the brain to reverify the exact approved scope and current external gates. Resume remains a separate request." if status == "brain_check_required" else
            "Prepare one bounded next scope and review its exact inheritance before approval. Pilot acceptance does not authorize more work." if meta["pilotPassed"] else
            "Select one bounded pilot scope, review its inheritance and approve its exact hashes. No blanket roadmap execution.",
        "brainPrompt": "Continue orchestration with $codex-orchestrator in this designated brain only. Start with status and read-only readiness/native project checks. Process only explicitly authorized control requests. Do not approve packets, resume dispatch, activate a heartbeat, create tasks or certify the pilot merely because this message was sent. Report the next exact approval or setup action needed.",
        "limits": "This diagnostic is not preflight or authorization. The brain must independently verify current packet/base/locks, predecessors, native setup, duplicate work, runner availability and repository policy at the creation boundary."}
